litmus-portal/cluster-agents/subscriber/pkg/workloads/workloads.go

```diff
@@ -0,0 +1,59 @@
+package workloads
+
+import (
+	"context"
+	"fmt"
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+func GetPodsFromWorkload(ns, kind string, allPods *corev1.PodList, wld []string, client *kubernetes.Clientset, getPodOwner func(ns, kind string, wld []string, client *kubernetes.Clientset) ([]string, string, error)) ([]string, error) {
+
+	ownerName, ownerKind, err := getPodOwner(ns, kind, wld, client)
+	if err != nil {
+		return nil, err
+	}
+
+	var pods []string
+	for _, r := range allPods.Items {
+		if checkOwnerRef(r.OwnerReferences, ownerName, ownerKind) {
+			pods = append(pods, r.Name)
+		}
+	}
+	return pods, nil
+}
+
+func GetPodsFromServices(ns string, wld []string, client *kubernetes.Clientset) ([]string, error) {
+	var pods []string
+	for _, svcName := range wld {
+		svc, err := client.CoreV1().Services(ns).Get(context.Background(), svcName, v1.GetOptions{})
+		if err != nil {
+			return nil, err
+		}
+		for k, v := range svc.Spec.Selector {
+			res, err := client.CoreV1().Pods(svc.Namespace).List(context.Background(), v1.ListOptions{LabelSelector: fmt.Sprintf("%s=%s", k, v)})
+			if err != nil {
+				return nil, err
+			}
+			for _, p := range res.Items {
+				pods = append(pods, p.Name)
+			}
```

Comment:
There could be possible overlaps for multiple selectors and as far as I remember services operate on  `and` operator for labels so all the labels should be present in the pod to be targeted by a service. In that case the pod listing should be done differently where all the labels are used.

Reply:
good catch! I didn't notice that.
Yes, we can directly use it to list all pods with matching selectors with & operator
```
pods, err := client.CoreV1().Pods(svc.Namespace).List(context.Background(), v1.ListOptions{LabelSelector: svc.Spec.Selector})
```